system/serial.c

```diff
@@ -101,13 +101,21 @@ void tty_init(void)
     unsigned char lcr;
 
     console_serial.enable       = true;
-    console_serial.is_mmio      = false;
+    console_serial.base_addr    = tty_address;
+    console_serial.baudrate     = tty_baud_rate;
     console_serial.parity       = SERIAL_DEFAULT_PARITY;
     console_serial.bits         = SERIAL_DEFAULT_BITS;
-    console_serial.baudrate     = tty_params_baud;
-    console_serial.reg_width    = 1;
-    console_serial.refclk       = UART_REF_CLK;
-    console_serial.base_addr    = serial_base_ports[tty_params_port];
+
+    // UART MMIO Address is usually above TOLUD and never < 1MB
+    if (console_serial.base_addr > 0xFFFF) {
+        console_serial.is_mmio      = true;
+        console_serial.reg_width    = 4;
```

Comment:
> Can you tell what is your board exactly? I'm aware of the linux blacklist (first link) but this is for very specific devices like Intel MID (Mobile Internet Device), basically Smartphone, tablets, etc.

It's Arista 5000 series, see [1]

I'm testing on one with Intel Xeon D-2700 series. As far as I can tell, Intel integrates this UART in their server-grade processors.

[1] https://www.arista.com/en/products/5000-series

> MMIO UART is not for the average Joe, so the user probably have the knowledge to change the code and compile it directly with the specific options needed on specific embedded platforms.

Pardon my ignorance, but how a user is supposed to now that their UART sits on MMIO? It sits on MMIO on most (if not all) of modern processors. ISA is dead and LPC is slowly dying. On modern AMD CPUs it's usually a local memory-mapped device, on Intel it's usually a PCI device.

> Base frequency may be detected and added later (not planned yet), but I'll put the value on a define.

You'll need an elaborate list of UARTs akin to one in linux. I'd recommend adding a command line argument to specify it. We already have to pass base address, and it's *a lot* easier to discover than base clock.

> Stride configuration will be added in the next commit (as mmio for 8 bit, mmio16 for 16 bit and mmio32 for 32bit)

I think it would be better to squash it into this one for bisectability purposes. Otherwise you'll have a different command-line interface between two adjacent commits.

Reply:
> Pardon my ignorance, but how a user is supposed to now that their UART sits on MMIO? 

A user not aware of what MMIO is usually doesn't know the required MMIO base address. A headless system with MMIO UART not linked to IPMI is usually not operated by a newbie.

> It sits on MMIO on most (if not all) of modern processors. ISA is dead and LPC is slowly dying. On modern AMD CPUs it's usually a local memory-mapped device, on Intel it's usually a PCI device.

You are likely thinking of servers/embedded users. I'm debugging console on a modern Asus PRIME Z790-A with a COM port still linked to 0x3f8 I/O right now (it doesn't support MMIO).

> You'll need an elaborate list of UARTs akin to one in linux. I'd recommend adding a command line argument to specify it. We already have to pass base address, and it's _a lot_ easier to discover than base clock.

I really don't want to spend time parsing a command line like `console=mmio32,0xFFDC0000,8n9600,13333333`. The very few people that really need this level of complexity can easily compile their own binary with all the tweak and very specific configuration they need.

> I think it would be better to squash it into this one for bisectability purposes. Otherwise you'll have a different command-line interface between two adjacent commits.

PR for a single feature are always squashed.

Reply:
PS : Of course, feel free to submit a PR with all the complex parsing code you think is needed. From my side, it's just that I don't have enough time to write code for all the elusive specific cases possible, nor the hardware required for debugging (all code paths must be tested on real hardware before merging).